broker/lib/fabrik/DirectorManager.js

```diff
@@ -177,7 +189,145 @@ class DirectorManager extends BaseManager {
     return this.director.getDeploymentIps(deploymentName);
   }
 
+  executePolicy(scheduled, action, deploymentName, dbUpdate, policyApplicable) {
+    if (!policyApplicable) {
+      return Promise.resolve(true);
+    }
+    if (dbUpdate) {
+      return Promise.resolve(true);
+    }
+    const targetDirectorConfig = this.director.getDirectorForOperation(action, deploymentName);
+    this.director.getCurrentTasks(action, targetDirectorConfig).then(out => {
+      let currentTasks, maxWorkers;
+      let allTasks = out.total;
+      let maxTasks = targetDirectorConfig.max_workers;
+      if (allTasks >= maxTasks) {
+        //no slots left anyway
+        return false;
+      }
+      if (scheduled) {
+        currentTasks = out.scheduled;
+        maxWorkers = targetDirectorConfig.policies.scheduled.max_workers;
+      } else {
+        currentTasks = out.user;
+        maxWorkers = targetDirectorConfig.policies.user.max_workers;
+      }
+      if (currentTasks < maxWorkers) {
+        return true;
+      }
+      return false;
+    });
+  }
+
+  getCurrentOperationState(serviceInstanceId) {
+    let output = {
+      'cached': false,
+      'task_id': null
+    };
+
+    return Promise.all([boshOperationCache.containsServiceInstance(serviceInstanceId), boshOperationCache.getBoshTask(serviceInstanceId)])
+      .spread((cached, taskId) => {
+        output.cached = cached;
+        output.task_id = taskId;
+      })
+      .return(output);
+  }
+
   createOrUpdateDeployment(deploymentName, params, args) {
+    logger.info(`Checking rate limits against bosh for deployment `);
+    const previousValues = _.get(params, 'previous_values');
+    const action = _.isPlainObject(previousValues) ? CONST.OPERATION_TYPE.UPDATE : CONST.OPERATION_TYPE.CREATE;
+    const scheduled = _.get(params, 'scheduled') || false;
+    const dbUpdate = _.get(params, '_serviceFabrikDbUpdate') || false;
+    _.omit(params, 'scheduled');
+    _.omit(params, '_serviceFabrikDbUpdate');
+
+    const decisionMaker = {
+      'shouldRunNow': false,
+      'cached': false,
+      'dbUpdate': false,
+      'policyApplicable': config.enable_bosh_rate_limit
+    };
+
+    return this.executePolicy(scheduled, action, deploymentName, dbUpdate, policyApplicable)
+      .then(shouldRunNow => {
+        if (!decisionMaker.policyApplicable) {
```

Comment:
Is this check again required since it was already checked in executePolicy method?

Reply:
This is needed because the `executePolicy` will return `runNow = true` but since the flag `policyApplicable` is responsible for the entire execution, it needs to be used subsequently